presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java

```diff
@@ -0,0 +1,311 @@
+        this.valueType = requireNonNull(valueType, "valueType is null");
+        this.keySet = keySet;
+        this.valueBlock = valueBlock;
+        this.mapBlock = mapBlock;
+        this.adder = getAdder(valueType);
+    }
+
+    public Type getKeyType()
+    {
+        return keyType;
+    }
+
+    public void serialize(BlockBuilder out)
+    {
+        assert mapBlock != null != (keySet != null && valueBlock != null);
+        BlockBuilder mapBlockBuilder = out.beginBlockEntry();
+        for (int i = 0; i < size(); i++) {
+            appendkey(i, mapBlockBuilder);
+            appendValue(i, mapBlockBuilder);
+        }
+        out.closeEntry();
+    }
+
+    private int size()
+    {
+        if (keySet != null) {
+            return keySet.size();
+        }
+
+        return mapBlock.getPositionCount() / 2;
+    }
+
+    private void appendkey(int i, BlockBuilder out)
+    {
+        if (mapBlock != null) {
+            keyType.appendTo(mapBlock, 2 * i, out);
+        }
+        else {
+            keyType.appendTo(keySet.getBlock(), i, out);
+        }
+    }
+
+    private void appendValue(int i, BlockBuilder blockBuilder)
+    {
+        if (mapBlock != null) {
+            appendValue(valueType, mapBlock, 2 * i + 1, blockBuilder);
+        }
+        else {
+            appendValue(valueType, valueBlock, i, blockBuilder);
+        }
+    }
+
+    private boolean isValueNull(int i)
+    {
+        if (mapBlock != null) {
+            return mapBlock.isNull(2 * i + 1);
+        }
+
+        return valueBlock.isNull(i);
+    }
+
+    private static void appendValue(Type valueType, Block block, int position, BlockBuilder blockBuilder)
+    {
+        if (block.isNull(position)) {
+            if (isExactNumericType(valueType)) {
+                valueType.writeLong(blockBuilder, 0L);
+            }
+            else {
+                valueType.writeDouble(blockBuilder, 0);
+            }
+        }
+        else {
+            valueType.appendTo(block, position, blockBuilder);
+        }
+    }
+
+    private int getPosition(TypedSet otherKeySet, int keyPosition)
+    {
+        if (mapBlock != null) {
+            return otherKeySet.positionOf(mapBlock, 2 * keyPosition);
+        }
+        else {
+            return otherKeySet.positionOf(keySet.getBlock(), keyPosition);
+        }
+    }
+
+    /**
+     * Only add this key value pair if we haven't seen this key before.
+     * Otherwise, ignore it.
+     */
+    public KeySetAndValues unionSum(KeySetAndValues other)
+    {
+        if (other.keySet != null &&
+                (keySet == null || keySet.size() < other.keySet.size())) {
+            // We can avoid create a new set by merging into the other one.
+            return other.unionSum(this);
+        }
+
+        // Just create the keyset
+        TypedSet resultKeySet = this.keySet;
+        if (resultKeySet == null) {
+            resultKeySet = new TypedSet(keyType, mapBlock.getPositionCount() / 2, "MAP_UNION_SUM");
```

Comment:
This means that both keySets are null right? When can this happen?

Reply:
Adding two single states. It should be clearer now with the two different classes. Just trying to do it in the class that already has a keyset.